### PR TITLE
connector testing: remove superfluous tasks

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Issue https://github.com/airbytehq/airbyte/issues/22130
This is a quick-win attempt at reducing test/build time for connectors.
Non-goals: investigate incremental build optimizations.

## How
What was wrong:
- Connector acceptance tests and base-normalization were built (and tested for CAT) on connector tests.

What I changed:
-  Use `latest` published *connector acceptance test* image on `/test` instead of building its image locally.  Developers can override this if they want to `/test` connector on their WIP CAT change by manually changing/committing [this line](https://github.com/airbytehq/airbyte/blob/ad2f39207bf0b5802476ee181205a9c1bb78559d/buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle#L8).
- Remove a legacy *hack* that I don't understand the purpose of. It runs connector-acceptance-test and base-normalization builds on connector test.

## 🚨 User Impact 🚨
* 🎉 **Speed gain**:  11mn faster on `source-pokeapi` ([PREVIOUSLY](https://github.com/airbytehq/airbyte/actions/runs/4183241096) vs [THIS PR CHANGE](https://github.com/airbytehq/airbyte/actions/runs/4205657035)). It's a 61% decrease in test duration... from 18mn to 7mn.

TODO: test this change on CDK, Java sources and Java destinations.

### CODE REVIEW

**Review by sherifnada — APPROVED** [review]
Great find!

File: buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle [hunk 1]
```diff
@@ -1,10 +1,27 @@
...
+        
         project.task('connectorAcceptanceTest') {
+            dependsOn project.build
+            dependsOn project.airbyteDocker
+            
             doFirst {
+                if (CONNECTOR_ACCEPTANCE_TEST_IMAGE != "dev") {
+                    project.exec {
+                        def args = [
+                            'docker', 'pull', 'airbyte/connector-acceptance-test:latest'
```
Comment (cpdeethree) [inline buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:20]: TODO: Perhaps highly consider using a versioned image here, since currently we can publish to this a breaking change which pulls the rug out from under our test suite. I'm ok with this going in as-is for now because the speed improvements are so dramatic

Comment (cpdeethree) [reply buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:20]: Is the base image static enough that we could go ahead and pin this today and you could remember to update it when you update the base image?

**Review by c-p-b — APPROVED** [review]
LGTM with one suggested change, either now or a TODO for later (unless you disagree, in which point we should have the convo ;) )

File: buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle (hunk 1 above)
Comment (alafanechere) [reply buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:20]: Yep, this is something I considered.
The image is quite dynamic but mainly maintained by an identified pool of engineer I trust enough to write unit tests and locally tests their changes before publishing.
I also like that we can quickly pin the version if the latest one has a bug.

There's also no better approach than waiting for nightly builds to run for all connectors on the latest version to assess the stability of tests. If the connector tests were reasonably fast we could automatically run the branch version of connector acceptance tests on a pool of previously succeeding connectors and then autobump the versions used on master if 🟢 . But this is not something we can currently achieve.

**Comment by alafanechere** [comment]
It looks like I've broken something on the java side according to the failing big query tests

File: buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle (hunk 1 above)
Comment (cpdeethree) [reply buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:20]: 👍 
If you agree, a TODO in the code is sufficient for me then. Otherwise feel free to leave as-is.

**Comment by alafanechere** [comment]
I think the error on the bigquery connector is on master too.

File: buildSrc/src/main/groovy/airbyte-python.gradle [hunk 2]
```diff
@@ -149,11 +149,6 @@ class AirbytePythonPlugin implements Plugin<Project> {
                 command = "install -r requirements.txt"
                 inputs.file('requirements.txt')
                 outputs.file('build/installedlocalreqs.txt')
-
-                // HACK: makes all integrations depend on installing requirements for bases. long term we should resolve deps and install in order.
-                if (project.getPath().startsWith(":airbyte-integrations:connectors") && !project.hasProperty("reports_folder")) {
-                    dependsOn project.rootProject.getTasksByName("airbytePythonApply", true).findAll { it.project.getPath().startsWith(":airbyte-integrations:bases") }
-                }
```
[the remark this answers is not in the document]
Comment (davinchia) [reply buildSrc/src/main/groovy/airbyte-python.gradle:156]: Sherif knows more than me here so I want him to confirm.

My thoughts:
I don't think we can remove this without changing more dependencies. This block basically makes each connector submodule depend on all the submodules in the bases folders e.g. (base, base-java, base-java-s3), this is required for connectors whose docker file depends on a `dev` base image. See [this](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-bigquery/Dockerfile#L1) for example.

To remove this, I think we would need to go through all the connector dockerfiles and make sure they either use a versioned base image so we don't have any local image expectations.

File: buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle [hunk 3]
```diff
@@ -1,10 +1,27 @@
 import org.gradle.api.Plugin
 import org.gradle.api.Project
 
+
 class AirbyteConnectorAcceptanceTestPlugin implements Plugin<Project> {
+    
+    // Change this value to 'dev' if you want to use a locally built connector acceptance test image
+    String CONNECTOR_ACCEPTANCE_TEST_IMAGE = 'latest'
```
Comment (pedroslopez) [inline buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:8]: Could this be exposed as an argument to the command somehow? I think this means when working on acceptance test changes, we won't be able to test our changes against connectors easily

Comment (evantahler) [reply buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:8]: An ENVIRONMENT variable might be a good way to handle this

Comment (alafanechere) [reply buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:8]: I've added an option to the `/test` command: `connector-acceptance-test-version`, which defaults to latest.
It's then passed to gradle as a project property via env var with `ORG_GRADLE_PROJECT_connectorAcceptanceTestVersion`.
So, on a Connector Acceptance Test PR, we can run `/test connector=connectors/source-pokeapi connector-acceptance-test-version=dev` to use a locally built CAT.

Comment (alafanechere) [reply buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:8]: Here's an example workflow to which I've passed connector-acceptance-test-version=dev : https://github.com/airbytehq/airbyte/actions/runs/4222475489

**Comment by alafanechere** [comment]
Re-requested review from @pedroslopez and @evantahler following my parameter addition to run a custom version (`dev`) of the connector acceptance test.

File: buildSrc/src/main/groovy/airbyte-python.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (cpdeethree) [reply buildSrc/src/main/groovy/airbyte-python.gradle:156]: What's the risk factor here of pushing ahead and addressing issues as they come up?

This is a massive improvement that will undoubtedly save us a not-insignificant amount of time and $$ on ci bills.

[the remark this answers is not in the document]
Comment (alafanechere) [reply buildSrc/src/main/groovy/airbyte-python.gradle:156]: @davinchia thanks for your reply.
I'm under the impression this "hack" is safe to remove because:
1. It's python only. As you mentioned, there could be more stickiness to `base` pre-installation on Java connector, but this change is python connector specific.
2. It's always formatting and installing base-normalization and connector-acceptance-test, even if these are not defined as a dependency on the connector side which I think is a superfluous task in the context of connector build.
3. We already have the guarantee that connector acceptance tests are installed in the connector because it is defined in their `requirement.txt` ([eg](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-pokeapi/requirements.txt#L1)) and installed in the `installLocalReqs` task.
I don't know exactly why we make connector-acceptance-test a test requirements dependency of connectors and why we declare this dependency both in `setup.py` and `requirements.txt`.

@sherifnada please share I you have more details on this matter 😄 🙏

File: buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle [hunk 4]
```diff
@@ -1,10 +1,35 @@
...
+    
     void apply(Project project) {
+        
         project.task('connectorAcceptanceTest') {
+            dependsOn project.build
+            dependsOn project.airbyteDocker
+            
+            if (!project.hasProperty('connectorAcceptanceTestVersion')) {
+                project.ext.connectorAcceptanceTestVersion = 'latest'
```
Comment (pedroslopez) [inline buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:17]: I'm missing some gradle knowledge here - sometimes I see this being referred as `project.ext.connectorAcceptanceTestVersion` and others as `project.connectorAcceptanceTestVersion`. What's the difference? Can they be used interchangeably?

I also see `CONNECTOR_ACCEPTANCE_TEST_IMAGE` being set to `latest` at the top and then checked on line 25 - should these usages be changed to refer to `project.connectorAcceptanceTestVersion` (or the `.ext` variant?)

Comment (pedroslopez) [inline buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:17]: I'm missing some gradle knowledge here - sometimes I see this being referred as `project.ext.connectorAcceptanceTestVersion` and others as `project.connectorAcceptanceTestVersion`. What's the difference? Can they be used interchangeably?

I also see `CONNECTOR_ACCEPTANCE_TEST_IMAGE` being set to `latest` at the top and then checked on line 25 - should these usages be changed to refer to `project.connectorAcceptanceTestVersion` (or the `.ext` variant?)

Comment (alafanechere) [reply buildSrc/src/main/groovy/airbyte-connector-acceptance-test.gradle:14]: @pedroslopez 

> I'm missing some gradle knowledge here - sometimes I see this being referred as project.ext.connectorAcceptanceTestVersion 

From what I understood: project properties are usually set directly in projects not it plugins (through env var, CLI input or declared in `build.gradle`). To set a project property from a plugin you've to use `project.ext.thePropertyIwantToset`. 

> I also see CONNECTOR_ACCEPTANCE_TEST_IMAGE being set to latest at the top and then checked on line 25 - should these usages be changed to refer to project.connectorAcceptanceTestVersion 

Yep sorry I forgot to remove its usage. it's done now.

File: buildSrc/src/main/groovy/airbyte-python.gradle (hunk 2 above)
[the remark this answers is not in the document]
Comment (davinchia) [reply buildSrc/src/main/groovy/airbyte-python.gradle:156]: @alafanechere I missed that this is only for the python modules. If that's true, I believe your comment is correct. I double checked by spot checking a couple of python connector build setups - don't seem like they directly rely on the various connector bases. 

Have you tried removing this and build 2 - 3 python connectors? That's the best sanity check.

**Review by pedroslopez — APPROVED** [review]
Code for connectorAcceptanceTestVersion LGTM!

Will let others comment on the removal of the "hack"

File: .github/workflows/test-command.yml [hunk 5]
```diff
@@ -111,6 +114,9 @@ jobs:
           ACTION_RUN_ID: ${{github.run_id}}
           # Oracle expects this variable to be set. Although usually present, this is not set by default on Github virtual runners.
           TZ: UTC
+          ORG_GRADLE_PROJECT_connectorAcceptanceTestVersion: ${{github.event.inputs.connector-acceptance-test-version}}
+          S3_BUILD_CACHE_ACCESS_KEY_ID: ${{ secrets.SELF_RUNNER_AWS_ACCESS_KEY_ID }}
+          S3_BUILD_CACHE_SECRET_KEY: ${{ secrets.SELF_RUNNER_AWS_SECRET_ACCESS_KEY }}
```
Comment (pedroslopez) [inline .github/workflows/test-command.yml:119]: why were these added?

Comment (alafanechere) [reply .github/workflows/test-command.yml:119]: It's an addition to allow gradle to use the remote cache stored on S3. It's something already setup on other gradle CI pipeline but was missing for this one.
